dali/pipeline/graph/op_graph2.cc

```diff
@@ -0,0 +1,313 @@
+bool OpGraph::EraseData(std::string_view name) {
+  auto it = name2data_.find(name);
+  if (it == name2data_.end())
+    return false;
+
+  auto &node = *it->second;
+  if (node.producer.op) {
+    assert(node.producer.op->outputs.size() >= static_cast<size_t>(node.producer.idx));
+    assert(node.producer.op->outputs[node.producer.idx] == &node);
+    node.producer.op->outputs[node.producer.idx] = nullptr;
+  }
+
+  for (auto &edge : node.consumers) {
+    if (edge.op == nullptr)
+      continue;
+    assert(edge.op->inputs.size() >= static_cast<size_t>(edge.idx));
+    assert(edge.op->inputs[edge.idx] == &node);
+    edge.op->inputs[edge.idx] = nullptr;
+  }
+
+  data_nodes_.erase(it->second->iter);
+  name2data_.erase(it);
+  return true;
+}
+
+int OpGraph::AddOutput(std::string_view name) {
+  auto it = name2data_.find(name);
+  if (it == name2data_.end())
+    throw std::invalid_argument(make_string(
+      "The name \"", name, "\" is not a name of a known DataNode."));
+  it->second->pipeline_output = true;
+  outputs_.push_back(it->second->name);
+  return outputs_.size() - 1;
+}
+
+/** Implements topological sorting via depth-first search */
+class OpGraph::SortHelper {
+ public:
+  explicit SortHelper(OpGraph &graph) : graph_(graph) {
+    sorted_ops_.reserve(graph_.op_nodes_.size());
+    sorted_data_.reserve(graph_.data_nodes_.size());
+  }
+
+  void Run() {
+    ClearVisitMarkers(graph_.op_nodes_);
+    ClearVisitMarkers(graph_.data_nodes_);
+
+    // First, go over the outputs
+    for (auto &data : graph_.data_nodes_) {
+      if (data.pipeline_output)
+        Traverse(&data);
+    }
+
+    // Then add operators marked as "keep".
+    for (auto &op : graph_.op_nodes_) {
+      if (op.keep)
+        Traverse(&op);
+    }
+
+    OpNodeList out_ops;
+    for (auto *op : sorted_ops_)
+      out_ops.splice(out_ops.end(), graph_.op_nodes_, op->iter);
+
+    DataNodeList out_data;
+    for (auto *data : sorted_data_)
+      out_data.splice(out_data.end(), graph_.data_nodes_, data->iter);
+
+    graph_.op_nodes_.swap(out_ops);
+    graph_.data_nodes_.swap(out_data);
+
+    // Now out_ops and out_data are the ops and data, respectively, that got removed.
+    // We should remove them from the name2xx maps and remove corresponding consumer links.
+    // NOTE: no producer of otherwise valid node should be removed, so we only need to adjust
+    // consumers.
+
+    for (auto &pruned_op : out_ops) {
+      graph_.RemoveDataNodeReferences(pruned_op);
+      graph_.name2op_.erase(pruned_op.instance_name);
+    }
+    for (auto &pruned_data : out_data) {
+      graph_.name2data_.erase(pruned_data.name);
+    }
+  }
+
+ private:
+  OpGraph &graph_;
+
+  std::vector<OpNode *> sorted_ops_;
+  std::vector<DataNode *> sorted_data_;
+
+  template <typename Node>
+  class Visit {
+   public:
+    explicit Visit(Node *n) : node_(n) {
+      if (node_->visit_pending)
+        throw std::logic_error("Cycle detected.");
+      node_->visit_pending = true;
```

Comment:
Do we clear `visit_pending` before entering the visiting phase? I only see that we clear `visited` markers, as there is no encapsulation, I think we should clear all at the start.

Reply:
`visit_pending` is cleared by the `Visit` helper. Perhaps I could move it to the header to make it available to other graph algorithms in the future.